### DISCUSSION TITLE
Scratch Project "Entry #1" – clones not working

### DISCUSSION THREAD

**Discussion by PullJosh** [leopard-js/leopard-mentors#4] [Cleanup Crew]
> _I am posting this [on behalf of Scratch user Dragon_Projects](https://scratch.mit.edu/users/PullJosh/#comments-336846367)._

### Scratch Project URL

https://scratch.mit.edu/projects/1040192471/

### What kind of error(s) are you experiencing?

- [ ] Leopard is failing to generate JavaScript code
- [X] The JavaScript code generated by Leopard doesn't work correctly
- [ ] The JavaScript code generated by Leopard is confusing or needs improvement
- [ ] Other

### Description of issue(s)

"I might have used it wrong, but it only shows one of the clones, the main text. It doesn't show any of the tiles in the background. _[...]_ The music and main text worked, but some of the clones didn't." - [comment](https://scratch.mit.edu/users/PullJosh/#comments-336840989)

Side note: "Oh yeah, another funny thing is that the Java[script] seemed to fix a bug in my scratch project, of the original text supposed to be coming in at 100% size. The beginning size works as I planned it to be on Java, but on Scratch it starts with a smaller size" - [comment](https://scratch.mit.edu/users/PullJosh/#comments-336846898)

**Comment by towerofnix** [comment]
We've started investigating this! We filed one pull request to help us debug the issue and catch similar ones in the future:

* leopard-js/leopard#199

We revived an old pull request that appears to more or less implement a fix, and started up the nicer implementation that we would rather merge instead, if it works out like we're hoping:

* leopard-js/sb-edit#149
* leopard-js/sb-edit#150

In summary, the tricky thing with this code is that some of the Scratch code sets the size of particular clones (for example, the tiles) to NaN. Leopard didn't handle that nicely at all before, and that caused errors, and for those clones not to be visible at all.

We're working on fixes to make Leopard's behavior more compatible with Scratch in this regard, but you can address it in your own Scratch or JavaScript code by taking a close look at any times you use the "set size to" block. (It's `this.size = ...` in JavaScript.) Look at the math blocks you are using carefully. If they might involve division by zero, for example, that would make the value NaN, and cause the trickiness we're seeing.

In Scratch, "set size to (NaN) %" - for example, "set size to (0 / 0) %" - is the same thing as "set size to 0%". This doesn't actually make the sprite infinitesimally tiny - it just makes it the "minimum" size that Scratch allows for the sprite's current costume. I could be mistaken, but I don't think "minimum size for a costume" (and "maximum size for a costume", for that matter) is a concept that exists in Leopard's code, yet. It's a point of incompatibility, where Leopard behaves differently from Scratch! And it's especially prominent here, because even if the code were doing `this.size = 0` like it's supposed to... well, I think that would result in the sprite being invisible, heh. Instead of just really tiny. So we have work to do, here, too!

We haven't carefully compared Leopard behavior (with a fix coded) and Scratch behavior for this project, yet, so there may be other things to fix, too. But the solution we tried so far looked pretty promising! Just some more Leopard code work, to get everything running right and nicely. ✨
